File: 2306/main.go
package main

import "fmt"

func main() {
	fmt.Println(distinctNames([]string{"abc", "bca"}))
}

func distinctNames(ideas []string) int64 {
	names := make(map[byte]map[string]struct{})
	for _, idea := range ideas {
		if _, exists := names[idea[0]]; !exists {
			names[idea[0]] = make(map[string]struct{})
		}
		names[idea[0]][idea[1:]] = struct{}{}
	}
	getIntersectSize := func(a, b map[string]struct{}) int {
		count := 0
		for key := range a {
			if _, found := b[key]; found {
				count++
			}
		}
		return count
	}
	var ans int64 = 0
	for preA, setA := range names {
		for preB, setB := range names {
			if preA == preB {
				continue
			}
			intersect := getIntersectSize(setA, setB)
			ans += int64(len(setA) - intersect) * int64(len(setB) - intersect)
		}
	}
	return ans
}

File: 2306/main.rs
use std::collections::{HashMap, HashSet};

struct Solution {}

impl Solution {
    pub fn distinct_names(ideas: Vec<String>) -> i64 {
        let mut names : HashMap<char, HashSet<String>> = HashMap::new();
        for idea in ideas.iter() {
            let first_char = idea.chars().next().unwrap();
            let suffix = idea[1..].to_string();
            names.entry(first_char).or_insert_with(HashSet::new).insert(suffix);
        }
        let get_intersect_size = |a: &HashSet<String>, b: &HashSet<String>| -> usize {
            a.intersection(b).count()
        };
        let mut ans: i64 = 0;
        for (pre_a, set_a) in names.iter() {
            for (pre_b, set_b) in names.iter() {
                if pre_a == pre_b {
                    continue
                }
                let intersect = get_intersect_size(set_a, set_b);
                ans += (set_a.len() - intersect) as i64 * (set_b.len() - intersect) as i64;
            }
        }
        ans
    }
}

fn main() {
    println!("{}", Solution::distinct_names(vec!["abc".to_string(), "bca".to_string()]));
}